### ISSUE TITLE
[JENKINS-45579] Step to set stage or parallel status

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#4315]
<p><b>Problem</b><br/>
There is no way to catch a failing block, mark it as "failed" with a user defined description and continue execution of the Pipeline. </p>

<p>In the example below, the <em>Optional Tests</em> stage would always be successful.</p>

<pre>
stage (<span class="code-quote">'Optional Tests'</span>) {
<span class="code-keyword">try</span> {
    echo <span class="code-quote">"Running optional tests..."</span>
    sh <span class="code-quote">'exit -1'</span>
} <span class="code-keyword">catch</span> (Exception err) {
    echo <span class="code-quote">'Optional tests failed... don'</span>t propagate failure'
}
}
</pre>

<p><b>Solution</b><br/>
Introduce new steps that can set the state of the stage to UNSTABLE, FAILED, SUCCESS, ABORTED. The message specified would be visible on the UI in Blue Ocean. </p>

<p><em>Example</em></p>

<pre>
stage (<span class="code-quote">'Optional Tests'</span>) {
<span class="code-keyword">try</span> {
    echo <span class="code-quote">"Running optional tests..."</span>
    sh <span class="code-quote">'exit -1'</span>
} <span class="code-keyword">catch</span> (Exception err) {
    unstable <span class="code-quote">'Optional tests failed... don'</span>t propagate failure'
}
}
</pre>

<p><b>Notes</b></p>
<ul>
	<li>There is precedence here with the <tt>error</tt> signal step.</li>
	<li>To ensure consistency the FAILED state should always fail the pipeline.</li>
	<li>The UNSTABLE state is used by tests to mark a stage as problematic but allows execution of the Pipeline to continue.</li>
	<li>Unstable should be expanded to handle the case where the intent of the Pipeline author is to record that there was a problem with the execution of the stage but execution is allowed to continue.</li>
</ul>

<p><b>Original Request</b><br/>
I recently discovered that pipeline stages that are run in parallel which report different results from one another are not currently being displayed correctly in the Blue Ocean UI. Here is a short snippet of code that reproduces the problem case I found:</p>

<pre>
node {
    parallel testA: {
stage (<span class="code-quote">"Required Tests"</span>) {
    echo <span class="code-quote">"Running required tests..."</span>
}
    }, testB: {
<span class="code-keyword">try</span> {
    stage (<span class="code-quote">"Optional Tests"</span>) {
echo <span class="code-quote">"Running optional tests..."</span>
sh <span class="code-quote">'exit -1'</span>
    }
} <span class="code-keyword">catch</span> (Exception err) {
    echo <span class="code-quote">"Optional tests failed... don't propagate failure"</span>
}
    }
    
    stage (<span class="code-quote">"Deploy"</span>) {
echo <span class="code-quote">"Deploying..."</span>
    }
}
</pre>
<p>Here I expect that 2 dummy test suites will be executed, one containing important tests that must all pass, and another containing less important tests that may be allowed to fail. Following these 2 parallel test stages is a final deploy stage that would be run afterwards so long as the "important" tests complete successfully.</p>

<p>In this example I'd expect that the node in the pipeline graph for the 'testB' stage would show up red while the other 2 nodes would show up green. Currently this does not happen - all 3 stages show up as green. I've attached a sample screenshot to illustrate the results.</p>

<p><span class="image-wrap" style=""><a id="38915_thumb" href="https://issues.jenkins.io/secure/attachment/38915/38915_Screen+Shot+2017-07-17+at+12.46.43+PM.png" title="Screen Shot 2017-07-17 at 12.46.43 PM.png" file-preview-type="image" file-preview-id="38915" file-preview-title="Screen Shot 2017-07-17 at 12.46.43 PM.png"><img src="https://issues.jenkins.io/secure/thumbnail/38915/_thumb_38915.png" style="border: 0px solid black" role="presentation"/></a></span></p>

<p>It is also worth mentioning that the statuses of these stages does appear to be represented correctly in the old 'stage view' from the default Jenkins UI. I've attached a screenshot illustrating this as well.</p>

<p><span class="image-wrap" style=""><a id="38914_thumb" href="https://issues.jenkins.io/secure/attachment/38914/38914_Screen+Shot+2017-07-17+at+12.46.53+PM.png" title="Screen Shot 2017-07-17 at 12.46.53 PM.png" file-preview-type="image" file-preview-id="38914" file-preview-title="Screen Shot 2017-07-17 at 12.46.53 PM.png"><img src="https://issues.jenkins.io/secure/thumbnail/38914/_thumb_38914.png" style="border: 0px solid black" role="presentation"/></a></span></p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=leedega">leedega</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579" target="_blank">Step to set stage or parallel status</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/dnusbaum.png" title="dnusbaum's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dnusbaum">dnusbaum</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Critical
<li><b>component(s)</b>: blueocean-plugin, pipeline
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2019-07-29T17:36:12+00:00
<li><b>votes</b>: 154
<li><b>watchers</b>: 173
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p><b>Problem</b>
There is no way to catch a failing block, mark it as "failed" with a user defined description and continue execution of the Pipeline. </p>

<p>In the example below, the <em>Optional Tests</em> stage would always be successful.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage (<span class="code-quote">'Optional Tests'</span>) {
        <span class="code-keyword">try</span> {
            echo <span class="code-quote">"Running optional tests..."</span>
            sh <span class="code-quote">'exit -1'</span>
        } <span class="code-keyword">catch</span> (Exception err) {
            echo <span class="code-quote">'Optional tests failed... don'</span>t propagate failure'
        }
}
</pre>
</div></div>

<p><b>Solution</b>
Introduce new steps that can set the state of the stage to UNSTABLE, FAILED, SUCCESS, ABORTED. The message specified would be visible on the UI in Blue Ocean. </p>

<p><em>Example</em></p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage (<span class="code-quote">'Optional Tests'</span>) {
        <span class="code-keyword">try</span> {
            echo <span class="code-quote">"Running optional tests..."</span>
            sh <span class="code-quote">'exit -1'</span>
        } <span class="code-keyword">catch</span> (Exception err) {
            unstable <span class="code-quote">'Optional tests failed... don'</span>t propagate failure'
        }
}
</pre>
</div></div>

<p><b>Notes</b></p>
<ul>
	<li>There is precedence here with the <tt>error</tt> signal step.</li>
	<li>To ensure consistency the FAILED state should always fail the pipeline.</li>
	<li>The UNSTABLE state is used by tests to mark a stage as problematic but allows execution of the Pipeline to continue.</li>
	<li>Unstable should be expanded to handle the case where the intent of the Pipeline author is to record that there was a problem with the execution of the stage but execution is allowed to continue.</li>
</ul>

<p><b>Original Request</b>
I recently discovered that pipeline stages that are run in parallel which report different results from one another are not currently being displayed correctly in the Blue Ocean UI. Here is a short snippet of code that reproduces the problem case I found:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">node {
    parallel testA: {
        stage (<span class="code-quote">"Required Tests"</span>) {
            echo <span class="code-quote">"Running required tests..."</span>
        }
    }, testB: {
        <span class="code-keyword">try</span> {
            stage (<span class="code-quote">"Optional Tests"</span>) {
                echo <span class="code-quote">"Running optional tests..."</span>
                sh <span class="code-quote">'exit -1'</span>
            }
        } <span class="code-keyword">catch</span> (Exception err) {
            echo <span class="code-quote">"Optional tests failed... don't propagate failure"</span>
        }
    }
    
    stage (<span class="code-quote">"Deploy"</span>) {
        echo <span class="code-quote">"Deploying..."</span>
    }
}
</pre>
</div></div>
<p>Here I expect that 2 dummy test suites will be executed, one containing important tests that must all pass, and another containing less important tests that may be allowed to fail. Following these 2 parallel test stages is a final deploy stage that would be run afterwards so long as the "important" tests complete successfully.</p>

<p>In this example I'd expect that the node in the pipeline graph for the 'testB' stage would show up red while the other 2 nodes would show up green. Currently this does not happen - all 3 stages show up as green. I've attached a sample screenshot to illustrate the results.</p>

<p><span class="image-wrap" style=""><a id="38915_thumb" href="https://issues.jenkins.io/secure/attachment/38915/38915_Screen+Shot+2017-07-17+at+12.46.43+PM.png" title="Screen Shot 2017-07-17 at 12.46.43 PM.png" file-preview-type="image" file-preview-id="38915" file-preview-title="Screen Shot 2017-07-17 at 12.46.43 PM.png"><img src="https://issues.jenkins.io/secure/thumbnail/38915/_thumb_38915.png" style="border: 0px solid black" role="presentation"/></a></span></p>

<p>It is also worth mentioning that the statuses of these stages does appear to be represented correctly in the old 'stage view' from the default Jenkins UI. I've attached a screenshot illustrating this as well.</p>

<p><span class="image-wrap" style=""><a id="38914_thumb" href="https://issues.jenkins.io/secure/attachment/38914/38914_Screen+Shot+2017-07-17+at+12.46.53+PM.png" title="Screen Shot 2017-07-17 at 12.46.53 PM.png" file-preview-type="image" file-preview-id="38914" file-preview-title="Screen Shot 2017-07-17 at 12.46.53 PM.png"><img src="https://issues.jenkins.io/secure/thumbnail/38914/_thumb_38914.png" style="border: 0px solid black" role="presentation"/></a></span></p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins LTS v2.46.3<br/>
Pipeline plugin v2.5<br/>
Blue Ocean v1.1.4
```
</details>
<details><summary><i>4 attachments</i></summary>

- [Screen Shot 2017-07-17 at 1.08.58 PM.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/38916/Screen%20Shot%202017-07-17%20at%201.08.58%20PM.png)
  > ![Screen Shot 2017-07-17 at 1.08.58 PM.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/38916/Screen%20Shot%202017-07-17%20at%201.08.58%20PM.png)
- [Screen Shot 2017-07-17 at 12.46.43 PM.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38915/Screen%20Shot%202017-07-17%20at%2012.46.43%20PM.png)
  > ![Screen Shot 2017-07-17 at 12.46.43 PM.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38915/Screen%20Shot%202017-07-17%20at%2012.46.43%20PM.png)
- [Screen Shot 2017-07-17 at 12.46.53 PM.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38914/Screen%20Shot%202017-07-17%20at%2012.46.53%20PM.png)
  > ![Screen Shot 2017-07-17 at 12.46.53 PM.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38914/Screen%20Shot%202017-07-17%20at%2012.46.53%20PM.png)
- [image-2019-04-12-18-20-27-978.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/46710/image-2019-04-12-18-20-27-978.png)
  > ![image-2019-04-12-18-20-27-978.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/46710/image-2019-04-12-18-20-27-978.png)
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote links associated with this issue:_


  - [CloudBees Internal CD-267](https://cloudbees.atlassian.net/browse/CD-267)
  - [CloudBees Internal CD-378](https://cloudbees.atlassian.net/browse/CD-378)
  - [jenkinsci/workflow-basic-steps-plugin#​83](https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/83)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>leedega:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=307227&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-307227">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>it may be worth noting that, when not using a parallel code block in my example above the Blue Ocean UI does correctly render the nodes in the appropriate color.</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/38916/38916_Screen+Shot+2017-07-17+at+1.08.58+PM.png" style="border: 0px solid black" /></span></p>
</pre>
</details>

</ul>
</details>

<p>it may be worth noting that, when not using a parallel code block in my example above the Blue Ocean UI does correctly render the nodes in the appropriate color.</p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/38916/Screen%20Shot%202017-07-17%20at%201.08.58%20PM.png" style="border: 0px solid black" /></span></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=307640&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-307640">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=leedega" class="user-hover" rel="leedega">leedega</a>,</p>

<p>Thanks for the report. I really appreciate the clear example demonstrating this.</p>

<p>Unfortunately, by catching the exception and eching, Blue Ocean has correctly interpreted the status of this stage as you have effectively eaten the error in the catch block. Not possible for us to determine the exact intent of your program automatically.</p>

<p>You'll have to find a different way to do this that causes the parallel to fail and bubble up this info. Commonly people set <tt>currentBuild.result = "UNSTABLE"</tt> when skipping over test failures. At the moment, that marks the whole pipeline as unstable but we are currently working on a way to set unstable for a single step/stage/parallel.</p>

<p>Thanks
James</p>
</pre>
</details>

</ul>
</details>

<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=leedega" class="user-hover" rel="leedega">leedega</a>,</p>

<p>Thanks for the report. I really appreciate the clear example demonstrating this.</p>

<p>Unfortunately, by catching the exception and eching, Blue Ocean has correctly interpreted the status of this stage as you have effectively eaten the error in the catch block. Not possible for us to determine the exact intent of your program automatically.</p>

<p>You'll have to find a different way to do this that causes the parallel to fail and bubble up this info. Commonly people set <tt>currentBuild.result = "UNSTABLE"</tt> when skipping over test failures. At the moment, that marks the whole pipeline as unstable but we are currently working on a way to set unstable for a single step/stage/parallel.</p>

<p>Thanks<br/>
James</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=307649&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-307649">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=leedega" class="user-hover" rel="leedega">leedega</a> I am discussing with our team internally how we might be able to solve this problem but in a slightly different way. Requires the issues that you already linked here. </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=leedega" class="user-hover" rel="leedega">leedega</a> I am discussing with our team internally how we might be able to solve this problem but in a slightly different way. Requires the issues that you already linked here. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=309988&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-309988">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=leedega" class="user-hover" rel="leedega">leedega</a> I've updated the description of this ticket to reflect a proposed solution. Please let me know if this looks good to you.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=leedega" class="user-hover" rel="leedega">leedega</a> I've updated the description of this ticket to reflect a proposed solution. Please let me know if this looks good to you.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mbadran.jpg" title="mbadran's avatar" /> mbadran:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=310805&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-310805">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a></p>

<p>Can you also please have that status extractable externally i.e. in a mailing groovy script. Usually the groovy scripts would use build result to show specific parts depending only on currentBuild.result causes problems when multiple parallel tasks tries to set the status just for the e-mail then return the status back to the original state.</p>

<p> </p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a></p>

<p>Can you also please have that status extractable externally i.e. in a mailing groovy script. Usually the groovy scripts would use build result to show specific parts depending only on currentBuild.result causes problems when multiple parallel tasks tries to set the status just for the e-mail then return the status back to the original state.</p>

<p> </p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/medianick.png" title="medianick's avatar" /> medianick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=328921&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-328921">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Given that <a href="https://issues.jenkins.io/browse/JENKINS-27092" title="create a step to abort the build with a NOT_BUILT status" class="issue-link" data-issue-key="JENKINS-27092">JENKINS-27092</a> has been closed with the intent to cover its needs on this one, I just wanted to add my need here: a simple way to abort a build while still reporting a successful status check back to GitHub (in a multibranch pipeline scenario). Right now I'm using a when condition on every stage to skip it (having detected a message like "<span class="error">&#91;ci skip&#93;</span>" in the git commit log).</p>
</pre>
</details>

</ul>
</details>

<p>Given that <a href="https://issues.jenkins.io/browse/JENKINS-27092" title="create a step to abort the build with a NOT_BUILT status" class="issue-link" data-issue-key="JENKINS-27092">JENKINS-27092</a> has been closed with the intent to cover its needs on this one, I just wanted to add my need here: a simple way to abort a build while still reporting a successful status check back to GitHub (in a multibranch pipeline scenario). Right now I'm using a when condition on every stage to skip it (having detected a message like "<span class="error">&#91;ci skip&#93;</span>" in the git commit log).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/felyppers.svg" title="felyppers's avatar" /> felyppers:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=331222&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-331222">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm having similar issues due to the unreadability of the pipeline, the Devs are complaining about it. Do we have any news on this feature of setting parallel steps individual results?</p>
</pre>
</details>

</ul>
</details>

<p>I'm having similar issues due to the unreadability of the pipeline, the Devs are complaining about it. Do we have any news on this feature of setting parallel steps individual results?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jekeller.png" title="jekeller's avatar" /> jekeller:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=331485&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-331485">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This is also somewhat related to steps like the warnings publisher, which if you run in parallel results in the whole build being marked unstable even though only a single segment should have been, which may not be fixed by making separate unstable results for each parallel step (unless it fully encapsulates the build?)</p>
</pre>
</details>

</ul>
</details>

<p>This is also somewhat related to steps like the warnings publisher, which if you run in parallel results in the whole build being marked unstable even though only a single segment should have been, which may not be fixed by making separate unstable results for each parallel step (unless it fully encapsulates the build?)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>progovoy:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=333904&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-333904">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Our Dev guys are also complaining about this issue. We would like to be able to mark certain stages as unstable / failed and still continue to execute the pipeline. Then we need the ability to check for any failures of the current run and not to execute the deploy stage. What do you think?</p>
</pre>
</details>

</ul>
</details>

<p>Our Dev guys are also complaining about this issue. We would like to be able to mark certain stages as unstable / failed and still continue to execute the pipeline. Then we need the ability to check for any failures of the current run and not to execute the deploy stage. What do you think?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>ass7:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=339037&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-339037">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>UP! <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/thumbs_up.png" height="16" width="16" align="absmiddle" alt="" border="0"/> I'd definitely be very happy to see that workint too though <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/star_yellow.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>UP! <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/thumbs_up.png" height="16" width="16" align="absmiddle" alt="" border="0"/> I'd definitely be very happy to see that workint too though <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/star_yellow.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nimrod.png" title="nimrod's avatar" /> nimrod:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=343599&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-343599">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>What's the status of the task?</p>

<p>It's strange that after a year+ of trying to switch our CI from <b>Build Flow</b> to <b>Pipelines/Blue Ocean</b>, I still find the the former more usable...</p>
</pre>
</details>

</ul>
</details>

<p>What's the status of the task?</p>

<p>It's strange that after a year+ of trying to switch our CI from <b>Build Flow</b> to <b>Pipelines/Blue Ocean</b>, I still find the the former more usable...</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>ekassis:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=344847&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-344847">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is there any progress on this issue?</p>
</pre>
</details>

</ul>
</details>

<p>Is there any progress on this issue?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sizgiyaev:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=356616&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-356616">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>HI, is there any update ? We really need that feature</p>
</pre>
</details>

</ul>
</details>

<p>HI, is there any update ? We really need that feature</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>pdecourcel:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=356967&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-356967">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi ! We really need this ! No news ?</p>
</pre>
</details>

</ul>
</details>

<p>Hi ! We really need this ! No news ?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>byahia:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=357240&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-357240">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'd love to see this implemented too. Is there any planned work to do it?</p>
</pre>
</details>

</ul>
</details>

<p>I'd love to see this implemented too. Is there any planned work to do it?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>shanidar1:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=357717&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-357717">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is there any update? </p>
</pre>
</details>

</ul>
</details>

<p>Is there any update? </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>idriver:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=363977&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-363977">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We also need this feature</p>
</pre>
</details>

</ul>
</details>

<p>We also need this feature</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=364592&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364592">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage (<span class="code-quote">'Optional Tests'</span>) {
  echo <span class="code-quote">'Running optional tests…'</span>
  <span class="code-keyword">if</span> (sh(script: <span class="code-quote">'exit -1'</span>, returnStatus: <span class="code-keyword">true</span>) != 0) {
    unstable <span class="code-quote">'Optional tests failed…<span class="code-keyword">do</span> not propagate failure'</span>
  }
}
</pre>
</div></div>
</pre>
</details>

</ul>
</details>


<pre>
stage (<span class="code-quote">'Optional Tests'</span>) {
  echo <span class="code-quote">'Running optional tests…'</span>
  <span class="code-keyword">if</span> (sh(script: <span class="code-quote">'exit -1'</span>, returnStatus: <span class="code-keyword">true</span>) != 0) {
    unstable <span class="code-quote">'Optional tests failed…<span class="code-keyword">do</span> not propagate failure'</span>
  }
}
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stuartrowe.png" title="stuartrowe's avatar" /> stuartrowe:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=364594&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364594">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> is <tt>unstable</tt> a step now? I thought it was only a condition for <tt>post</tt> in declarative pipelines.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> is <tt>unstable</tt> a step now? I thought it was only a condition for <tt>post</tt> in declarative pipelines.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=364604&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364604">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>No, sorry for confusion. I was just sketching an edit to the original example by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=leedega" class="user-hover" rel="leedega">leedega</a> to not rely on a <tt>try</tt>-<tt>catch</tt> block, which would be undesirable in this context as it would catch not just nonzero exit status from a script but all sorts of other errors and even build interrupts and timeouts.</p>
</pre>
</details>

</ul>
</details>

<p>No, sorry for confusion. I was just sketching an edit to the original example by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=leedega" class="user-hover" rel="leedega">leedega</a> to not rely on a <tt>try</tt>-<tt>catch</tt> block, which would be undesirable in this context as it would catch not just nonzero exit status from a script but all sorts of other errors and even build interrupts and timeouts.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>saad_azam:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=364832&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364832">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Another thing is that the behavior described earlier about the fact that build stage shows the correct node rendering when not using parallel (link to the comment below) does not hold true for a declarative pipeline.</p>

<p><a href="https://issues.jenkins-ci.org/browse/JENKINS-45579?focusedCommentId=307227&amp;page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-307227" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-45579?focusedCommentId=307227&amp;page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-307227</a></p>
</pre>
</details>

</ul>
</details>

<p>Another thing is that the behavior described earlier about the fact that build stage shows the correct node rendering when not using parallel (link to the comment below) does not hold true for a declarative pipeline.</p>

<p><a href="https://issues.jenkins-ci.org/browse/JENKINS-45579?focusedCommentId=307227&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-307227" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-45579?focusedCommentId=307227&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-307227</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nofarblue:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=365438&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-365438">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hello, </p>

<p>I'm a Product Manager at CloudBees and we're currently working on a design for this request, and I would love to get your feedback on the suggested solution:</p>

<p>We were thinking of expanding the functionality of "<a href="#catcherror-catch-error-and-set-build-result" target="_blank" rel="noopener"><b>catchError</b></a>]" block as follows:</p>

<p>Current default behavior (will be maintained for backward compatibility) when an error is caught:</p>
<ul>
	<li>Step result is failed (though execution of the pipeline continues)</li>
	<li>Stage result is success</li>
	<li>Build result is failed</li>
</ul>


<p><ins>Suggested changes:</ins></p>

<p>Add ability to set arguments to determine the desired results upon a caught error:</p>
<ul>
	<li>Stage result : success/failure (default: success)</li>
	<li>Build result: success/failure (default: failure)</li>
</ul>


<p> </p>

<p>In addition, a very common use case is the ability to set a warning - mark the build and a stage in it as one that warrants attention without failing the build. To support that, we're considering adding a new block called "<b>warnError</b>", which will behave as follows:</p>
<ul>
	<li>If an error is caught:
	<ul>
		<li>Step result is failed</li>
		<li>Stage result is set to unstable</li>
		<li>Build result is set to unstable</li>
	</ul>
	</li>
</ul>


<p>(we're also fixing the issue where if a build becomes unstable it's hard to pinpoint which stage caused the build to become unstable )</p>

<p> </p>

<p>Thank you, </p>

<p>Nofar Bluestein,</p>

<p>Product Manager, CloudBees</p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>Hello, </p>

<p>I'm a Product Manager at CloudBees and we're currently working on a design for this request, and I would love to get your feedback on the suggested solution:</p>

<p>We were thinking of expanding the functionality of "<a href="#catcherror-catch-error-and-set-build-result" target="_blank" rel="noopener"><b>catchError</b></a>]" block as follows:</p>

<p>Current default behavior (will be maintained for backward compatibility) when an error is caught:</p>
<ul>
	<li>Step result is failed (though execution of the pipeline continues)</li>
	<li>Stage result is success</li>
	<li>Build result is failed</li>
</ul>


<p><ins>Suggested changes:</ins></p>

<p>Add ability to set arguments to determine the desired results upon a caught error:</p>
<ul>
	<li>Stage result : success/failure (default: success)</li>
	<li>Build result: success/failure (default: failure)</li>
</ul>


<p> </p>

<p>In addition, a very common use case is the ability to set a warning - mark the build and a stage in it as one that warrants attention without failing the build. To support that, we're considering adding a new block called "<b>warnError</b>", which will behave as follows:</p>
<ul>
	<li>If an error is caught:
	<ul>
		<li>Step result is failed</li>
		<li>Stage result is set to unstable</li>
		<li>Build result is set to unstable</li>
	</ul>
	</li>
</ul>


<p>(we're also fixing the issue where if a build becomes unstable it's hard to pinpoint which stage caused the build to become unstable )</p>

<p> </p>

<p>Thank you, </p>

<p>Nofar Bluestein,</p>

<p>Product Manager, CloudBees</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>brianjenkins:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=365443&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-365443">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nofarblue" class="user-hover" rel="nofarblue">nofarblue</a> - Sounds great. I'm not too concerned about syntax decisions, as long as the basic issues are addressed. This sounds like a very reasonable solution. Glad the issue is getting some attention!</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nofarblue" class="user-hover" rel="nofarblue">nofarblue</a> - Sounds great. I'm not too concerned about syntax decisions, as long as the basic issues are addressed. This sounds like a very reasonable solution. Glad the issue is getting some attention!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>saad_azam:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=365495&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-365495">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nofarblue" class="user-hover" rel="nofarblue">nofarblue</a> - Many thanks for the response. The changes you suggested are exactly what I am looking forward to have in declarative pipeline. In case it is not too much of an ask for you, it would be nice to get a rough estimate about when are the suggested changes going to be rolled out. </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nofarblue" class="user-hover" rel="nofarblue">nofarblue</a> - Many thanks for the response. The changes you suggested are exactly what I am looking forward to have in declarative pipeline. In case it is not too much of an ask for you, it would be nice to get a rough estimate about when are the suggested changes going to be rolled out. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nancyrobertson2005:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=365513&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-365513">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nofarblue" class="user-hover" rel="nofarblue">nofarblue</a> - Many thanks for looking into this.  Your proposal sounds good.  I am particularly interested in the handling of the unstable condition.  Our pipelines are mostly scripted, and we have been using the error() function to mark failed stages, but it would help to be able to mark stages as unstable and keep going.  Does your current plan include the addition of an unstable() step?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nofarblue" class="user-hover" rel="nofarblue">nofarblue</a> - Many thanks for looking into this.  Your proposal sounds good.  I am particularly interested in the handling of the unstable condition.  Our pipelines are mostly scripted, and we have been using the error() function to mark failed stages, but it would help to be able to mark stages as unstable and keep going.  Does your current plan include the addition of an unstable() step?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nofarblue:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=365629&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-365629">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thank you for the feedback!</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=saad_azam" class="user-hover" rel="saad_azam">saad_azam</a> I cannot commit to an ETA at the moment, but I can tell you it's a high priority for us,. I will provide additional update soon. </p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nancyrobertson2005" class="user-hover" rel="nancyrobertson2005">nancyrobertson2005</a> in scripted pipelines the following should work: </p>

<p> `warnError { error() }` would have the same result, but I will consider adding 'unstable' step so it's more elegant and consistent with the 'error' step</p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>Thank you for the feedback!</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=saad_azam" class="user-hover" rel="saad_azam">saad_azam</a> I cannot commit to an ETA at the moment, but I can tell you it's a high priority for us,. I will provide additional update soon. </p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nancyrobertson2005" class="user-hover" rel="nancyrobertson2005">nancyrobertson2005</a> in scripted pipelines the following should work: </p>

<p> `warnError { error() }` would have the same result, but I will consider adding 'unstable' step so it's more elegant and consistent with the 'error' step</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>shanidar1:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=365695&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-365695">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nofarblue" class="user-hover" rel="nofarblue">nofarblue</a> Thanks for the update! looking forward to use the stage status solution.  </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nofarblue" class="user-hover" rel="nofarblue">nofarblue</a> Thanks for the update! looking forward to use the stage status solution.  </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/dnusbaum.png" title="dnusbaum's avatar" /> dnusbaum:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=367038&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-367038">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Pipeline: Basic Steps Plugin version 2.16 added the new <tt>warnError</tt> and <tt>unstable</tt> steps and changes to <tt>catchError</tt> that <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nofarblue" class="user-hover" rel="nofarblue">nofarblue</a> mentioned, see the <a href="https://wiki.jenkins.io/display/JENKINS/Pipeline+Basic+Steps+Plugin?#PipelineBasicStepsPlugin-2.16(May14,2019)" class="external-link" target="_blank" rel="nofollow noopener">2.16 release notes</a> for details on the specifics. See <a href="https://issues.jenkins-ci.org/browse/JENKINS-39203?focusedCommentId=367029&amp;page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-367029" class="external-link" target="_blank" rel="nofollow noopener">this comment on JENKINS-39203</a> for additional information on related issues. You need to update to Pipeline: Graph Analysis 1.10 for these steps to be visualized correctly by Blue Ocean.</p>

<p>Should we close this issue and open more specific issues related to the request for a step to set the build result to <tt>FAILURE</tt> but continue execution and a step to set the build result to <tt>ABORTED</tt> and abort the build, or handle them as part of this ticket?</p>
</pre>
</details>

</ul>
</details>

<p>Pipeline: Basic Steps Plugin version 2.16 added the new <tt>warnError</tt> and <tt>unstable</tt> steps and changes to <tt>catchError</tt> that <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nofarblue" class="user-hover" rel="nofarblue">nofarblue</a> mentioned, see the <a href="https://wiki.jenkins.io/display/JENKINS/Pipeline+Basic+Steps+Plugin?#PipelineBasicStepsPlugin-2.16(May14,2019)" class="external-link" target="_blank" rel="nofollow noopener">2.16 release notes</a> for details on the specifics. See <a href="https://issues.jenkins-ci.org/browse/JENKINS-39203?focusedCommentId=367029&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-367029" class="external-link" target="_blank" rel="nofollow noopener">this comment on JENKINS-39203</a> for additional information on related issues. You need to update to Pipeline: Graph Analysis 1.10 for these steps to be visualized correctly by Blue Ocean.</p>

<p>Should we close this issue and open more specific issues related to the request for a step to set the build result to <tt>FAILURE</tt> but continue execution and a step to set the build result to <tt>ABORTED</tt> and abort the build, or handle them as part of this ticket?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rrrrrrrr:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=367209&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-367209">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi. Would you mind adding an example in <a href="https://github.com/jenkinsci/pipeline-examples" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/pipeline-examples</a> ? I'm not exactly an expert and coming up with the right syntax  assembling bit of code &amp; release note is quite hard.</p>
</pre>
</details>

</ul>
</details>

<p>Hi. Would you mind adding an example in <a href="https://github.com/jenkinsci/pipeline-examples" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/pipeline-examples</a> ? I'm not exactly an expert and coming up with the right syntax  assembling bit of code & release note is quite hard.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>saad_azam:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=367374&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-367374">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dnusbaum" class="user-hover" rel="dnusbaum">dnusbaum</a> I have tried the latest update with newly added 'unstable' step as well as updated 'catchError'. I am now able to setup and visualize individual stage status independently. Many thanks.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dnusbaum" class="user-hover" rel="dnusbaum">dnusbaum</a> I have tried the latest update with newly added 'unstable' step as well as updated 'catchError'. I am now able to setup and visualize individual stage status independently. Many thanks.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/dnusbaum.png" title="dnusbaum's avatar" /> dnusbaum:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=367618&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-367618">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rrrrrrrr" class="user-hover" rel="rrrrrrrr">rrrrrrrr</a> You can use the "Pipeline Syntax" link in the sidebar of a Pipeline job in Jenkins to get a visual interface that will show you the options provided by <tt>catchError</tt>, <tt>warnError</tt>, and <tt>unstable</tt> and will allow you to generate the actual syntax that would go in a Pipeline given the options you specify.</p>

<p>I will look into adding something to the examples repo.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rrrrrrrr" class="user-hover" rel="rrrrrrrr">rrrrrrrr</a> You can use the "Pipeline Syntax" link in the sidebar of a Pipeline job in Jenkins to get a visual interface that will show you the options provided by <tt>catchError</tt>, <tt>warnError</tt>, and <tt>unstable</tt> and will allow you to generate the actual syntax that would go in a Pipeline given the options you specify.</p>

<p>I will look into adding something to the examples repo.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>brianjmurrell:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=368375&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-368375">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I used the Pipeline Syntax snippet generator to generate me a <tt>catchError</tt> step and it produced:</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>catchError(buildResult: 'SUCCESS', stageResult: 'FAILURE') {
 // some block
}
</pre>
</div></div>

<p>which then chokes Jenkins with:</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>WorkflowScript: 147: Expecting "class hudson.model.Result" for parameter "buildResult" but got "SUCCESS" of type class java.lang.String instead @ line 147, column 49.
         catchError(buildResult: 'SUCCESS',
                                 ^
</pre>
</div></div>

<p>Removing the single quotes surrounding the <tt>SUCCESS</tt> and <tt>FAILURE</tt> values seems to at least keep the linter happy and the pipeline at least runs.  Still waiting to see if these new values have the desired effect though.  <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>I used the Pipeline Syntax snippet generator to generate me a <tt>catchError</tt> step and it produced:</p>



```
catchError(buildResult: 'SUCCESS', stageResult: 'FAILURE') {
 // some block
}

```

<p>which then chokes Jenkins with:</p>



```
WorkflowScript: 147: Expecting "class hudson.model.Result" for parameter "buildResult" but got "SUCCESS" of type class java.lang.String instead @ line 147, column 49.
 catchError(buildResult: 'SUCCESS',
 ^

```

<p>Removing the single quotes surrounding the <tt>SUCCESS</tt> and <tt>FAILURE</tt> values seems to at least keep the linter happy and the pipeline at least runs.  Still waiting to see if these new values have the desired effect though.  <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/dnusbaum.png" title="dnusbaum's avatar" /> dnusbaum:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=368376&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-368376">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=brianjmurrell" class="user-hover" rel="brianjmurrell">brianjmurrell</a> Yes, turns out that Declarative doesn't like the parameter types for `catchError`. See <a href="https://issues.jenkins.io/browse/JENKINS-57537" title="catchError  - Expecting &quot;class hudson.model.Result&quot; for parameter &quot;buildResult&quot;" class="issue-link" data-issue-key="JENKINS-57537"><del>JENKINS-57537</del></a> for more information, I am working on a fix that should be ready soon.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=brianjmurrell" class="user-hover" rel="brianjmurrell">brianjmurrell</a> Yes, turns out that Declarative doesn't like the parameter types for `catchError`. See <a href="https://issues.jenkins.io/browse/JENKINS-57537" title="catchError  - Expecting "class hudson.model.Result" for parameter "buildResult"" class="issue-link" data-issue-key="JENKINS-57537"><del>JENKINS-57537</del></a> for more information, I am working on a fix that should be ready soon.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>brianjmurrell:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=368384&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-368384">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dnusbaum" class="user-hover" rel="dnusbaum">dnusbaum</a> Oh, this is a pity.  <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/sad.png" height="16" width="16" align="absmiddle" alt="" border="0"/>  <tt>catchError(buildResult: 'SUCCESS', stageResult: 'FAILURE')</tt> was going to solve exactly the problem I newly have on hand today, which is how to let a step fail, mark it's stage as failed but not fail the overall job.</p>

<p>I'm really loathed to make the step and stage look like it succeeded (because it will mask failures that need investigating) just so that the rest of the stages get run and overall job passes.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dnusbaum" class="user-hover" rel="dnusbaum">dnusbaum</a> Oh, this is a pity.  <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/sad.png" height="16" width="16" align="absmiddle" alt="" border="0"/>  <tt>catchError(buildResult: 'SUCCESS', stageResult: 'FAILURE')</tt> was going to solve exactly the problem I newly have on hand today, which is how to let a step fail, mark it's stage as failed but not fail the overall job.</p>

<p>I'm really loathed to make the step and stage look like it succeeded (because it will mask failures that need investigating) just so that the rest of the stages get run and overall job passes.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>brianjmurrell:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=368405&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-368405">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dnusbaum" class="user-hover" rel="dnusbaum">dnusbaum</a> New issue filed about <tt>catchError()</tt>, and the <tt>post</tt> processing block, FWIW.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dnusbaum" class="user-hover" rel="dnusbaum">dnusbaum</a> New issue filed about <tt>catchError()</tt>, and the <tt>post</tt> processing block, FWIW.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/dnusbaum.png" title="dnusbaum's avatar" /> dnusbaum:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45579?focusedId=372322&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-372322">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The original example in the description of this ticket was fixed as part of <a href="https://issues.jenkins.io/browse/JENKINS-39203" title="All stages show up as UNSTABLE when only one stage should" class="issue-link" data-issue-key="JENKINS-39203"><del>JENKINS-39203</del></a> with the new <tt>warnError</tt> and <tt>unstable</tt> steps and changes to the <tt>catchError</tt> step, so I am going to close this issue See <a href="https://jenkins.io/blog/2019/07/05/jenkins-pipeline-stage-result-visualization-improvements/" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.io/blog/2019/07/05/jenkins-pipeline-stage-result-visualization-improvements/</a> and <a href="https://issues.jenkins-ci.org/browse/JENKINS-45579?focusedCommentId=367038&amp;page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-367038" class="external-link" target="_blank" rel="nofollow noopener">this earlier comment</a> for details. if you have a request for a step that is related to the general title of the ticket but is not covered by the changes I mentioned, please file a new ticket along the lines of <a href="https://issues.jenkins.io/browse/JENKINS-27092" title="create a step to abort the build with a NOT_BUILT status" class="issue-link" data-issue-key="JENKINS-27092">JENKINS-27092</a>.</p>
</pre>
</details>

</ul>
</details>

<p>The original example in the description of this ticket was fixed as part of <a href="https://issues.jenkins.io/browse/JENKINS-39203" title="All stages show up as UNSTABLE when only one stage should" class="issue-link" data-issue-key="JENKINS-39203"><del>JENKINS-39203</del></a> with the new <tt>warnError</tt> and <tt>unstable</tt> steps and changes to the <tt>catchError</tt> step, so I am going to close this issue See <a href="https://jenkins.io/blog/2019/07/05/jenkins-pipeline-stage-result-visualization-improvements/" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.io/blog/2019/07/05/jenkins-pipeline-stage-result-visualization-improvements/</a> and <a href="https://issues.jenkins-ci.org/browse/JENKINS-45579?focusedCommentId=367038&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-367038" class="external-link" target="_blank" rel="nofollow noopener">this earlier comment</a> for details. if you have a request for a step that is related to the general title of the ticket but is not covered by the changes I mentioned, please file a new ticket along the lines of <a href="https://issues.jenkins.io/browse/JENKINS-27092" title="create a step to abort the build with a NOT_BUILT status" class="issue-link" data-issue-key="JENKINS-27092">JENKINS-27092</a>.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-27092%22">JENKINS-27092</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-49764%22">JENKINS-49764</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `epic` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-45818%22">JENKINS-45818</a>]</i>
